app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt

```diff
@@ -229,56 +218,46 @@ class PoTokenWebView private constructor(
         val poToken = try {
             u8ToBase64(poTokenU8)
         } catch (t: Throwable) {
-            popPoTokenEmitter(identifier)?.onError(t)
+            popPoTokenContinuation(identifier)?.resumeWithException(t)
             return
         }
 
         if (BuildConfig.DEBUG) {
             Log.d(TAG, "Generated poToken: identifier=$identifier poToken=$poToken")
         }
-        popPoTokenEmitter(identifier)?.onSuccess(poToken)
+        popPoTokenContinuation(identifier)?.resume(poToken)
     }
 
-    override fun isExpired(): Boolean {
-        return Instant.now().isAfter(expirationInstant)
-    }
+    override val isExpired get() = Instant.now() > expirationInstant
     //endregion
 
     //region Handling multiple emitters
     /**
-     * Adds the ([identifier], [emitter]) pair to the [poTokenEmitters] list. This makes it so that
-     * multiple poToken requests can be generated invparallel, and the results will be notified to
-     * the right emitters.
+     * Adds the ([identifier], [continuation]) pair to the [poTokenContinuations] list. This makes
+     * it so that multiple poToken requests can be generated in parallel, and the results will be
+     * notified to the right continuations.
      */
-    private fun addPoTokenEmitter(identifier: String, emitter: SingleEmitter<String>) {
-        synchronized(poTokenEmitters) {
-            poTokenEmitters.add(Pair(identifier, emitter))
-        }
+    private fun addPoTokenEmitter(identifier: String, continuation: Continuation<String>) {
+        poTokenContinuations[identifier] = continuation
     }
```

Comment:
So as far as I understand, `addPoTokenEmitter` is called exactly once inside `generatePoToken`.

`generatePoToken` is called exactly twice, both times in `getWebClientPoToken`. And here’s the kicker: in the same `suspend` block, right underneath each other.

So this leads me to believe that the whole infrastructure of having a list of continuations is not really necessary, because everything happens one after the other anyway? I mean yes, in theory they could run at the same time, but in practice there’s only two calls running, and in sequence.

Reply:
What I’m saying here is that rewrite to continuations all well and good, but I think we don’t even need such a complicated setup in the first place.